feat(TypeAheadSelect): Add TypeAheadSelect component with storybook and tests

Laviro opened the pull request:
[**Storybook**](https://rawgit.com/Laviro/patternfly-react/storybook/feature/auto-complete/index.html?knob-Clear%20Button=true&knob-Multiple%20Selection=true&selectedKind=patternfly-react%2FForms%20and%20Controls%2FTypeAhead%20Select&selectedStory=Multiple%20Selection&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs) - Updated !

re #190
re #281

jeff-phillips-18 commented:
@patternfly/patternfly-react-ux

serenamarie125 commented:
@jeff-phillips-18 this is awaiting my input, sorry :(  I've started and will reply tomorrow noon at latest
Should we pull of css review until ux is approved ?

dabeng commented:
1. It will be better that key words are identified with bold font
<img width="326" alt="screen shot 2018-06-04 at 10 46 06 am" src="https://user-images.githubusercontent.com/5689075/40896150-08c2ab2e-67e6-11e8-903f-a3725005f379.png">

2. It will be better that allowing users to select multiple items in search box
<img width="411" alt="screen shot 2018-06-04 at 10 46 50 am" src="https://user-images.githubusercontent.com/5689075/40896245-756bd052-67e6-11e8-9c27-2f9006a64117.png">

Laviro commented:
Thanks for the comment @dabeng, 

1. I have added now a component that highlights the matching search term,
which is passed to the AutoComplete `renderMenu prop`.
![highlightresults](https://user-images.githubusercontent.com/26363699/40920550-963ad286-6815-11e8-9527-69ae375a0107.png)


You can check the [Highlight Results example](https://rawgit.com/Laviro/patternfly-react/storybook/feature/auto-complete/index.html?knob-highlight%20Only%20Result=true&knob-select%20Hint%20On%20Enter=true&knob-Allow%20Adding%20New%20Items=true&selectedKind=patternfly-react%2FWidgets%2FAutoComplete&selectedStory=Highlight%20Results&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook-addon-viewport%2Faddon-panel) 

You think we should make it as the default behavior? Or just provide the component to developers who want to use it?

2. As for the multiple items, 
in my opinion the options which are already shown in the search pane are what the *user selected*. 
The dark blue color that appears there is only for showing you which option you are on right now, 
in case you are going to delete it.

serenamarie125 commented:
@Laviro I'd like to suggest the following visual recommendations from @jennyhaines
<img width="902" alt="screen shot 2018-06-07 at 4 01 48 pm" src="https://user-images.githubusercontent.com/12504403/41123336-4f2c4a06-6a6c-11e8-945c-f1d216fede70.png">

serenamarie125 commented:
@Laviro we also suggest that the pattern name be "TypeAhead Select" and be in the "Forms and Controls" area

jennyhaines commented:
@Laviro, Also the text color of "No matches found." seems to use the #9c9c9c coloring which isn't AA accessibility compliant. I suggest that the color of this text be pf-black-600 (#72767b) so it can be more easily read, but still looks much lighter than regular dark text.

Here's what it looks like now:
<img width="766" alt="nomatchesfound" src="https://user-images.githubusercontent.com/24898297/41123927-4af3874a-6a6e-11e8-8f8c-7625a31d16d5.png">

terezanovotna commented:
@serenamarie125 @jennyhaines Can we add this visual guide for chips to PF? It would be great to have it documented. I've seen a lot of chips implementations lately and it has always been a question

jeff-phillips-18 commented on packages/core/less/autocomplete.less:
```diff
@@ -0,0 +1,246 @@
+    -moz-box-shadow: inset 0 1px 1px rgba(0, 0, 0, 0.075), 0 0 6px #67b168;
+    box-shadow: inset 0 1px 1px rgba(0, 0, 0, 0.075), 0 0 6px #67b168;
+  }
+}
```
We should import the css from react-bootstrap-typeahead, then we need only include what we want to override here.

jeff-phillips-18 commented on packages/core/src/components/AutoComplete/AutoComplete.js:
```diff
@@ -0,0 +1,46 @@
+  placeholder: null,
+  selectHintOnEnter: false
+};
+export default AutoComplete;
```
For both AutroComplete and AsyncAutoComplete, do we want to try to keep up with the props defined by react-bootstrap-typeahead? We could just use a pass thru and point at their documentation. Doesn't give a lot of information in our storybook but maybe pointing to their documentation is enough.

jennyhaines commented:
@terezanovotna - that makes sense to me! I'll add a visual story to the PF backlog and reference this  PR!

terezanovotna commented:
@jennyhaines thanks so much 👏👏👏

Laviro commented:
@serenamarie125 @jennyhaines Thank you for the very detailed review, I fixed those things and later Jenny and I agreed that the chip should be a bit darker, and set it to `pf-black-200`.

@jeff-phillips-18 Thank you for the code review, 
now the `bootstrap-react-typeahead` styles are imported in our sass/less where we override with our styles. I deleted the proptypes that are well documented in the `bootstrap-react-typeahead` library's github page and we'll link there.

jeff-phillips-18 commented on packages/core/less/autocomplete.less:
```diff
@@ -0,0 +1,44 @@
+  color: #363636;
+}
+
+button.close.rbt-close.rbt-token-remove-button {
```
To override this, we just need `.rbt-token .rbt-token-remove-button`

jeff-phillips-18 commented on packages/core/less/autocomplete.less:
```diff
@@ -0,0 +1,44 @@
+@import '~react-bootstrap-typeahead/css/Typeahead.css';
+
+div.rbt-token.rbt-token-removeable {
```
Can we remove the `div.`?

jeff-phillips-18 commented on packages/core/less/autocomplete.less:
```diff
@@ -0,0 +1,44 @@
+  background-color: @color-pf-black-150;
+  height: 20px;
+  font-size: 12px;
+  color: #363636;
```
This should be a color from the patternfly color palette using a `color-pf-` variable, like color-pf-black-900

Laviro replied:
In the css review they ask specifically for this color, 
though I think it can be replaced with `color-pf-black-800`

jeff-phillips-18 replied:
Ah, ok. Visual designer comments override mine here 😉

jeff-phillips-18 commented on packages/core/less/autocomplete.less:
```diff
@@ -0,0 +1,44 @@
+}
+
+button.close.rbt-close.rbt-token-remove-button {
+  color: #363636;
```
color variable

jeff-phillips-18 commented on packages/core/less/autocomplete.less:
```diff
@@ -0,0 +1,44 @@
+  margin-top: 0px;
+}
+
+a.dropdown-item.disabled {
```
Can we add some specificity here so as not to effect other components?

jeff-phillips-18 commented on packages/core/less/autocomplete.less:
```diff
@@ -0,0 +1,44 @@
+div.form-control.rbt-input-multi.rbt-input {
+  height: auto;
+}
+
```
Could we just use:
```
.rbt-input-multi {
  &.rbt-input-wrapper {
    ...
   }
  &.rbt-input {
    ...
  }
}
```

jeff-phillips-18 commented on packages/core/less/autocomplete.less:
```diff
@@ -0,0 +1,44 @@
+  height: auto;
+}
+
+input.rbt-input-main {
```
do we need the `.input`?

jeff-phillips-18 commented on packages/core/less/autocomplete.less:
```diff
@@ -0,0 +1,44 @@
+  padding: 2px 6px;
+}
+
+ul.rbt-menu.dropdown-menu.show {
```
do we need the `.ul`?

serenamarie125 commented:
@Laviro is it possible to override the color of the "chips" ? The reason I'm asking is "IF" this component is used to select "tags" or "labels", i wonder if we would want the ability to change the visualization of the chip to look like that tag/label visualization.  We don't necessarily need to do this as part of the PR, but bringing it up in case we want this initial contribution to be able to support this type of change in the future

Laviro commented:
@serenamarie125 Thanks for the approve !!
Yeah, the original `react-bootstrap-typeahead` uses blue style for their "chips",
here we override with grey style and anyone who will want to change their color is welcome to do so :)

serenamarie125 commented:
@Laviro one last suggestion 😄 

for the case when the user is adding a new entry, I'm wondering rather than saying "New selection: **xxx**" if it could say "Create and select item: **xxx**"

<img width="964" alt="screen shot 2018-06-12 at 8 19 49 am" src="https://user-images.githubusercontent.com/12504403/41300615-29b48f42-6e34-11e8-9dab-6e37ce1bc3f4.png">

jeff-phillips-18 commented:
The text for that can be set by the application via props. Maybe for our example (storybook) we can use @serenamarie125's preferred text, but know that an application will need to set the text or they will get the default `New selection:`

Laviro commented:
@serenamarie125 Done ! the only thing needed to change is the `newSelectionPrefix` prop and anyone can set their own message. I have set it to "Create and select item: "

jeff-phillips-18 commented:
Could you use `yarn commit` to correctly format the commit message?

Rohoover commented:
All - I was looking at this component because we may have a use for it in Satellite and I had some UX related questions and comments.

1) **Multiple Selection:** When I added more items than the field could allow, it seemed to hide all the new selections while continuing to add my new selections. I had no way of knowing what my old selections were. Is there a way to see my other selections?

2) When making a single selection only, is it intended that the currently selected item then become the only item in the drop down? (See behavior in Menu Alignment, Input Size & Highlight results)

jeff-phillips-18 commented:
@Rohoover 

1)  looks like an issue. Could you add it please?

2) Since this is TypeAhead Select not just a dropdown, only the matches are shown in single select mode. If you backspace you start seeing additional matches.

jeff-phillips-18 commented:
If you want to file an issue on 2) with expected behavior/results we can look at it there. Seems like that would just be a simple dropdown behavior though. But I agree with the `x` or some method to clear the selection.
